### PR TITLE
Reland "Upgrade template Gradle, App AGP, Module AGP, and Kotlin versions, and tests"... but no longer upgrade module AGP version

### PR DESCRIPTION
**Pull request by gmackall**
Relands https://github.com/flutter/flutter/pull/150969, but removes the upgrade to the module AGP version.

The reason is that a more complicated change is required because in AGP 8.0 software components are no longer generated by default, but rather only generated for variants that are configured to be published using the publishing DSL (see the `android.disableAutomaticComponentCreation` section of https://developer.android.com/build/releases/past-releases/agp-8-0-0-release-notes).

That broke our aar initialization script, because the components didn't exist so the `aar` tasks never got [created here](https://github.com/flutter/flutter/blob/9ff9c67272f7f95c7d53ac3f89e0b659df9ddf3e/packages/flutter_tools/gradle/aar_init_script.gradle#L29).

Verified that the one postsubmit that failed now passes (`android_obfuscate_test`)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl [hunk 1]
```diff
@@ -30,6 +30,20 @@ group = "{{androidIdentifier}}"
...
+        }
+        singleVariant("profile") {
+            withSourcesJar()
+            withJavadocJar()
+        }
+        singleVariant("release") {
+            withSourcesJar()
+            withJavadocJar()
+        }
+    }
```
Comment (gmackall) [inline packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:46]: If any reviewer is knowledgeable about Maven publishing, I'd like to know if this will have unintended consequences. It looks like we were already modifying the publishing in the aar_init script as well https://github.com/flutter/flutter/blob/9ff9c67272f7f95c7d53ac3f89e0b659df9ddf3e/packages/flutter_tools/gradle/aar_init_script.gradle#L38, so it seems similar to stuff we are already doing. But I don't understand it very well

Comment (reidbaker) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:46]: I have some experience with maven but not with using the publishing gradle task. 
One thing I notice is that this hardcodes the variants and maybe that is ok for a template but if so we should update the documentation for adding variants to including modifying this.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl [hunk 2]
```diff
@@ -30,6 +30,20 @@ group = "{{androidIdentifier}}"
 version = "1.0"
 
 android {
+    publishing {
```
Comment (reidbaker) [inline packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:33]: Add a comment above this 
```
// publishing comes from gradle plugin 'maven-publish' 
// https://docs.gradle.org/current/userguide/publishing_maven.html
```

At least this is where I think publishing comes from.

Comment (reidbaker) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:33]: I guess I dont see that plugin being added so maybe it comes from https://developer.android.com/reference/tools/gradle-api/8.3/null/com/android/build/api/dsl/Publishing?

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl [hunk 3]
```diff
@@ -30,6 +30,20 @@ group = "{{androidIdentifier}}"
...
+        singleVariant("debug") {
```
Comment (reidbaker) [inline packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:34]: My gut says we should either use multiVariant or setup only release to publish and not this version where we use single variant multiple times. 
https://developer.android.com/reference/tools/gradle-api/8.3/null/com/android/build/api/dsl/LibraryPublishing

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl
Comment (reidbaker) [inline packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:1]: Can you comment on why you wanted/needed to add this publishing step and the existing flutter build aar was not good enough?

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 4]
```diff
@@ -3759,7 +3759,7 @@ void main() {
       logger.clear();
     }
   }, overrides: <Type, Generator>{
-    Java: () => FakeJava(version: const software.Version.withText(14, 0, 0, '14.0.0')), // Middle compatible Java version with current template AGP/Gradle versions.
+    Java: () => FakeJava(version: const software.Version.withText(20, 0, 0, '20.0.0')), // Middle compatible Java version with current template AGP/Gradle versions.
```
Comment (reidbaker) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3762]: Why did this need to change? I think 14 should still pass for "middle"

File: packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart [hunk 5]
```diff
@@ -913,10 +913,10 @@ A problem occurred evaluating project ':app'.
           '│ To fix this issue, replace the following content:                                │\n'
           '│ /android/build.gradle:                                                           │\n'
           "│     - classpath 'com.android.tools.build:gradle:<current-version>'               │\n"
-          "│     + classpath 'com.android.tools.build:gradle:7.3.0'                           │\n"
+          "│     + classpath 'com.android.tools.build:gradle:$templateAndroidGradlePluginVersion'                           │\n"
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart:916]: ~This may seem silly but when I open this file in github neither of these variables exist when I search. Are you sure this works?~ 
Edit: consider named import instead.

File: packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart [hunk 6]
```diff
@@ -530,7 +531,7 @@ void main() {
         Event.flutterBuildInfo(
           label: 'gradle-expected-file-not-found',
           buildType: 'gradle',
-          settings: 'androidGradlePluginVersion: 7.6.3, fileExtension: .aab',
+          settings: 'androidGradlePluginVersion: $templateDefaultGradleVersion, fileExtension: .aab',
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart:534]: Consider using a named import to make it easier to track where this variable comes from. 
`import 'package:flutter_tools/src/android/gradle_utils.dart' as gradleUtils;`

File: packages/flutter_tools/test/general.shard/android/gradle_test.dart [hunk 7]
```diff
@@ -546,9 +546,9 @@ flutter:
     });
   });
 
-  group('gradgradle_utils.le version', () {
+  group('gradle version', () {
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_test.dart:549]: I genuinely wonder how I messed up 3 of these. sorry!

File: packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart [hunk 8]
```diff
@@ -913,10 +913,10 @@ A problem occurred evaluating project ':app'.
           '│ To fix this issue, replace the following content:                                │\n'
           '│ /android/build.gradle:                                                           │\n'
           "│     - classpath 'com.android.tools.build:gradle:<current-version>'               │\n"
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart:915]: These 2 lines have double quote instead of single quote? can you fix them while you are here?

File: packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart [hunk 9]
```diff
@@ -779,7 +779,7 @@ dependencies {
                 gradleV: data.gradleVersion,
                 agpV: data.agpVersion,
               ),
-              data.validPair ? isTrue : isFalse,
+              data.validPair,
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:782]: Why this change? isTrue I think gives a better error message than comparing to true.

File: packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart [hunk 10]
```diff
@@ -967,19 +967,21 @@ allprojects {
...
+      JavaAgpTestData(false, javaVersion: '17', agpVersion: '8.5'),
       JavaAgpTestData(true, javaVersion: '17', agpVersion: maxKnownAndSupportedAgpVersion),
+      JavaAgpTestData(true, javaVersion: '17', agpVersion: '8.3'),
       JavaAgpTestData(true, javaVersion: '17', agpVersion: '8.1'),
       JavaAgpTestData(true, javaVersion: '17', agpVersion: '8.0'),
       JavaAgpTestData(true, javaVersion: '17', agpVersion: '7.4'),
-      JavaAgpTestData(false, javaVersion: '17.0.3', agpVersion: '8.2'),
+      JavaAgpTestData(false, javaVersion: '17.0.3', agpVersion: '8.5'),
       JavaAgpTestData(true, javaVersion: '17.0.3', agpVersion: maxKnownAndSupportedAgpVersion),
+      JavaAgpTestData(true, javaVersion: '17.0.3', agpVersion: '8.3'),
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:984]: Should we add a test for 8.2 here?

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 2 above)
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:33]: Yes, I think `project.publishing` would refer the the gradle plugin `maven-publish`, but that `project.android.publishing` is just a closure that exists in the Android Gradle plugin.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:1]: The existing `flutter build aar` doesn't work after the upgrade to the AGP version.

This is because of this change:

> AGP 8.0 creates no SoftwareComponent by default. Instead AGP creates SoftwareComponents only for variants that are configured to be published using the publishing DSL.

We were relying on the existence of the `release`, `profile`, and `debug` `SoftwareComponent`s to add our [`assembleAar$variant` task](https://github.com/flutter/flutter/blob/9ff9c67272f7f95c7d53ac3f89e0b659df9ddf3e/packages/flutter_tools/gradle/aar_init_script.gradle#L89), which we rely on as part of `flutter build aar`.

So adding the publishing step ensures that we start making the relevant `SoftwareComponent`s again.

File: packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart (hunk 8 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart:915]: From what I can tell this was actually done intentionally, because the line itself contains single quotes:
```
classpath 'com.android.tools.build:gradle:<current-version>'
```

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart (hunk 4 above)
Comment (gmackall) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3762]: This test is testing that the (middle Java among those compatible with the templates) works, where the lower bound is set by AGP (previously 11, but now 17 after the change to AGP 8.1) and the upper bound is set by Gradle. So 14 is out of the range (below)

File: packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart (hunk 9 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:782]: Ah I think in my skimming I read this as a case of `boolean ? true : false`, my mistake. I'll change back

File: packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart (hunk 10 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:984]: I've added one, but the tests here will never be exhaustive regardless because it is really only a test of the unique combination of Java `17.0.3` and AGP `8.2`, not a test of AGP `8.2` independently

File: packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart (hunk 6 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart:534]: Hmm I can change this if you have a strong preference, but it will require many other changes, e.g. every call to `parseGradleVersionFromDistributionUrl`, `getAgpVersion`, `getGradleVersion`, etc.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 1 above)
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:46]: I think people actually can't modify this file - it is one of the "ephemeral" add to app gradle files that gets re-generated frequently.

File: packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart (hunk 8 above)
Comment (reidbaker) [reply packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart:915]: ah that makes sense I didnt catch that.

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart (hunk 4 above)
Comment (reidbaker) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3762]: Sg.

File: packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart (hunk 6 above)
Comment (reidbaker) [reply packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart:534]: ok maybe leave it alone in this file and add it in the files where you added an import.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 1 above)
Comment (reidbaker) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:46]: If this gets generated then we need to support custom variants.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl
Comment (reidbaker) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:1]: Thank you this makes total sense. I would actually encourage you to add that a comment in the doc. It takes a pretty deep understanding to know those are related and I think we should leave ourselves breadcrumbs.

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 3 above)
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl]: Resolving as this is now only relevant to follow up pr

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl:1]: Resolving as this is now only relevant to follow up pr

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 2 above)
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl]: Resolving as this is now only relevant to follow up pr

File: packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl (hunk 1 above)
Comment (gmackall) [reply packages/flutter_tools/templates/module/android/library_new_embedding/Flutter.tmpl/build.gradle.tmpl]: Resolving as this is now only relevant to follow up pr

File: packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart (hunk 5 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart:916]: Done in the file where it is a new import

File: packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart (hunk 6 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart]: Done!

File: packages/flutter_tools/test/general.shard/android/gradle_test.dart (hunk 7 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/gradle_test.dart:549]: It looks like it was probably a find and replace thing, and no worries considering I probably missed it in review :)

**Review by reidbaker — APPROVED** [review]

**Comment by gmackall** [comment]
cc @christopherfujino no action needed, just an fyi that the change to the tests that we discussed in the last pr re-landed (because without a cc the last update you would have seen was the revert)
